collect_app/src/main/java/org/odk/collect/android/tasks/ChosenFileSaver.java

```diff
@@ -0,0 +1,105 @@
+package org.odk.collect.android.tasks;
+
+import android.content.Context;
+import android.net.Uri;
+import android.provider.MediaStore;
+
+import org.odk.collect.android.R;
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.dao.helpers.ContentResolverHelper;
+import org.odk.collect.android.exception.GDriveConnectionException;
+import org.odk.collect.android.logic.FormController;
+import org.odk.collect.android.utilities.FileUtils;
+import org.odk.collect.android.utilities.MediaUtils;
+
+import java.io.File;
+import java.lang.ref.WeakReference;
+
+import io.reactivex.Observable;
+import timber.log.Timber;
+
+public class ChosenFileSaver {
+
+    private final WeakReference weakReferenceContext;
+    private final Uri selectedFile;
+    private File newFile;
+    private int errorMessageRes;
+
+    private ChosenFileSaver(Context context, Uri selectedFile) {
+        weakReferenceContext = new WeakReference<>(context);
+        this.selectedFile = selectedFile;
+    }
+
+    public static Observable<SaveResult> createObservable(Context context, Uri selectedFile) {
```

Comment:
There are several `Observable` types that can be used for different situations like these 😄 The base `Observable` type is more fitted for a stream of elements. In this case, we are dealing with a single file so we can use the `Single<T>` type 😎 This type basically states that when this `Observable` executes then it should emit a single value or throw an error.  If it's a case where the file can be null, then we can opt to use the `Maybe<T>` that would say, if there's a possibility there can be a file not then it can be used. I was thinking of using this, but so far I just thought if there's an issue that would be thrown via an exception so I didn't.

Reply:
Thanks for the great advice! :+1:

Reply:
No prob!